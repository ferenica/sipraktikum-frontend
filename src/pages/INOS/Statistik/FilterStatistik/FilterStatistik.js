import React, {Component} from 'react';
import {withStyles} from '@material-ui/core/styles';
import Select from '@material-ui/core/Select';
import FormControl from '@material-ui/core/FormControl';
import MenuItem from '@material-ui/core/MenuItem';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Remove';
import '../StatistikPage.css';

const useStyles = (theme) => ({
  root: {
    paddingLeft: theme.spacing(3), paddingRight: theme.spacing(3),
    paddingTop: theme.spacing(3), paddingBottom: theme.spacing(10),
    display: 'flex',
    flexDirection: 'column',
  },
  formLabel: {
    fontSize: 18,
    fontWeight: 'bold',
    marginTop: '4vh',
    marginBottom: '1.1vh',
  },
  formControl: {
    // backgroundColor: "#F5F6F8",
    backgroundColor: 'white',
    color: '#FF9C51',
    borderColor: '#FF9C51',
    marginBottom: '1.1vh',
  },
  tambahButton: {
    color: 'rgb(255, 156, 81)',
    fontWeight: 'bold',
    borderColor: '#FF9C51',
  },
});
/**
 *
 */
class FilterStatistik extends Component {
  /**
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);
    this.nowYear = new Date().getFullYear();
    this.BATAS_TAHUN_AWAL = 2019;

    this.BATAS_TOTAL_TAHUN = 3;
  }

  /**
   *
   */
  UNSAFE_componentWillMount() {
    const years = [];
    for (let i=this.nowYear; i>=this.BATAS_TAHUN_AWAL; i--) {
      years.push(i);
    }
    this.setState({listTahun: years, chosenTahun: [this.nowYear]});
  }


  state = {
    category: 'Jenis Institusi',
    listTahun: [], chosenTahun: [],
    formTahunIdx: [],
  }

  handleChange_category = (event) => {
    this.setState({category: event.target.value}, () => {
      this.props.kategoriStatistik(this.state.category);
    });
  };

  handleChange_tahun = (event, tahunindex) => {
    const tempChosen = this.state.chosenTahun.slice();
    tempChosen[tahunindex] = event.target.value;
    this.setState({chosenTahun: tempChosen}, () => {
      this.props.tahunStatistik(this.state.chosenTahun);
    });
  };

  newTahunForm = (_event, index) => {
    // let newTahunVal = this.nowYear;
    // while (this.state.chosenTahun.includes(newTahunVal)) { newTahunVal-- }
    this.setState({
      chosenTahun: [...this.state.chosenTahun, ''],
      formTahunIdx: [...this.state.formTahunIdx, index],
    }, () => this.props.tahunStatistik(this.state.chosenTahun));
  }

  delTahunForm = (_event, index) => {
    this.setState({
      chosenTahun: this.state.chosenTahun.filter((tahun) =>
        tahun !== this.state.chosenTahun[index]),
      formTahunIdx: this.state.formTahunIdx.filter((tahunIdx) =>
        tahunIdx !== this.state.formTahunIdx[index-1]),
    }, () => this.props.tahunStatistik(this.state.chosenTahun));
  }

  /**
   *
   */
  render() {
    const {classes} = this.props;

    const tahunItems = this.state.listTahun.map((aYear) => {
      if (this.state.chosenTahun.includes(aYear)) {
        return <MenuItem disabled key={aYear} value={aYear}>{aYear}</MenuItem>;
      }
      return <MenuItem key={aYear} value={aYear}>{aYear}</MenuItem>;
    });

    return (
      <div className={classes.root}>
        <div className={classes.formLabel}>Berdasarkan</div>
        <FormControl variant="outlined" className={classes.formControl}>
          <Select id="select"
            value={this.state.category}
            onChange={this.handleChange_category}
            displayEmpty
            className= {classes.selectEmpty}
            inputProps={{'aria-label': 'Without label'}}
          >
            {/* daftar pilihan filter */}
            <MenuItem value={'Jenis Institusi'}>Jenis Institusi</MenuItem>
            <MenuItem value={'Tema'}>Tema</MenuItem>
          </Select>
        </FormControl>

        <div className={classes.formLabel}>Tahun</div>
        <FormControl variant="outlined" className={classes.formControl}>
          <Select id="select"
            value={this.state.chosenTahun[0]}
            onChange={(e) => this.handleChange_tahun(e, 0)}
            displayEmpty
            className={classes.selectEmpty}
            inputProps={{'aria-label': 'Without label'}}
          >
            {/* daftar pilihan filter */}
            {tahunItems}
          </Select>
        </FormControl>

        {/* tambahan filter Tahun */}
        {this.state.formTahunIdx.map((idxTahun) => {
          return (
            <FormControl variant="outlined" className={classes.formControl} key={idxTahun}>
              <Select id="selectTahun"
                value={this.state.chosenTahun[idxTahun]}
                onChange={(e) => this.handleChange_tahun(e, idxTahun)}
                displayEmpty
                className={classes.selectEmpty}
                inputProps={{'aria-label': 'Without label'}}
              >
                {/* daftar pilihan filter */}
                <MenuItem disabled value={''}>Pilih tahun..</MenuItem>
                {tahunItems}
              </Select>
            </FormControl>
          );
        })}

        <Grid container spacing={1}>
          <Grid className="MuiGrid-root MuiGrid-item btnGrid" style={{justifyContent: 'start'}}>
            {(this.state.chosenTahun.length > 1) &&
            <Button className={classes.tambahButton} id="KurangiBtn"
              onClick={(e) => this.delTahunForm(e, this.state.chosenTahun.length-1)}
            >
              <RemoveIcon />
            </Button>
            }
          </Grid>
          <Grid className="MuiGrid-root MuiGrid-item emptyGrid"/>
          <Grid className="MuiGrid-root MuiGrid-item btnGrid">
            {(this.state.chosenTahun.length < this.BATAS_TOTAL_TAHUN) &&
            <Button className={classes.tambahButton} id="TambahBtn"
              onClick={(e) => this.newTahunForm(e, this.state.chosenTahun.length)}
            >
              <AddIcon />
            </Button>
            }
          </Grid>
        </Grid>

      </div>
    );
  }
}

export default withStyles(useStyles)(FilterStatistik);
