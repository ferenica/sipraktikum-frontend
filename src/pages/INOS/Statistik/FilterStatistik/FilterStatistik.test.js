import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import FilterStatistik from './FilterStatistik.js';

Enzyme.configure({adapter: new EnzymeAdapter()});

describe('FilterStatistik', () => {
  const wrapper = shallow(<FilterStatistik
    kategoriStatistik={jest.fn()}
    tahunStatistik={jest.fn()}
  />);
  const divedWrapper = wrapper.dive();

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should run handleChange when an option is chosen', () => {
    const changeCategEvent = {target: {value: 'Tema'}};
    const changeTahunEvent = {target: {value: 2020}};

    divedWrapper.find('#select').at(0).simulate('change', changeCategEvent);
    divedWrapper.find('#select').at(1).simulate('change', changeTahunEvent);
    expect(divedWrapper.state().category).toBe('Tema');
    expect(divedWrapper.state().chosenTahun[0]).toBe(2020);
  });

  it('should add new Tahun form when Tambah button is clicked', () => {
    divedWrapper.find('#TambahBtn').simulate('click');
    expect(divedWrapper.state().chosenTahun.length).toBe(2);

    divedWrapper.find('#selectTahun').at(0).simulate('change', {target: {value: 2017}});
    expect(divedWrapper.state().chosenTahun[1]).toBe(2017);
  });

  it('should remove Tambah Button when 3 Tahun forms exists', () => {
    // there were 2 forms already, so only one more click event needed
    divedWrapper.find('#TambahBtn').simulate('click');

    expect(divedWrapper.state().chosenTahun.length).toBe(3);
    expect(divedWrapper.find('#TambahBtn').length).toBe(0);
  });

  it('should remove a Tahun form when Kurangi button is clicked', () => {
    divedWrapper.find('#KurangiBtn').simulate('click');

    // there were 3 forms before
    expect(divedWrapper.state().chosenTahun.length).toBe(2);
  });

  it('should not have Kurangi button when only 1 Tahun forms exist', () => {
    // there were 2 forms before
    divedWrapper.find('#KurangiBtn').simulate('click');
    expect(divedWrapper.state().chosenTahun.length).toBe(1);

    expect(divedWrapper.find('#selectTahun').length).toEqual(0);
    expect(divedWrapper.find('#KurangiBtn').length).toEqual(0);
  });
});
