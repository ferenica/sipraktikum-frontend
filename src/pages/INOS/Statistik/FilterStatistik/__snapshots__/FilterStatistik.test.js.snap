// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterStatistik should renders correctly 1`] = `ShallowWrapper {}`;
